File: src/cpp/IdpPacket.cpp
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for
// full license information.
#include "IdpPacket.h"

IdpPacket::IdpPacket (uint32_t payloadLength, IdpFlags flags, uint16_t source,
                      uint16_t destination, bool sealed)
{
    _writeIndex = 0;
    _readIndex = 0;
    _isSealed = false;

    auto length = 11 + payloadLength;

    if (((uint8_t) flags & (uint8_t) IdpFlags::CRC) == (uint8_t) IdpFlags::CRC)
    {
        length += 4;
    }

    _buffer = std::shared_ptr<uint8_t> (new uint8_t[length],
                                        std::default_delete<uint8_t[]> ());

    Write ((uint8_t) 0x02);
    Write (length);
    Write ((uint8_t) flags);
    Write (source);
    Write (destination);

    _isSealed = sealed;
}

IdpFlags IdpPacket::Flags ()
{
    return (IdpFlags) Data ()[5];
}

uint16_t IdpPacket::Source ()
{
    uint16_t result;

    memcpy (&result, (const void*) (Data () + 6), sizeof (uint16_t));

    if (BitConverter::IsLittleEndian ())
    {
        result = BitConverter::SwapEndian (result);
    }

    return result;
}

uint16_t IdpPacket::Destination ()
{
    uint16_t result;

    memcpy (&result, (const void*) (Data () + 8), sizeof (uint16_t));

    if (BitConverter::IsLittleEndian ())
    {
        result = BitConverter::SwapEndian (result);
    }

    return result;
}

void IdpPacket::Seal ()
{
    Write ((uint8_t) 0x03);

    if (((uint8_t) Flags () & (uint8_t) IdpFlags::CRC) ==
        (uint8_t) IdpFlags::CRC)
    {
        // TODO write CRC.
        Write ((uint32_t) 0xAA55AA55);
    }

    _isSealed = true;
}

void IdpPacket::Write (const void* data, uint32_t length)
{
    memcpy (_buffer.get () + _writeIndex, data, length);

    _writeIndex += length;
}

uint8_t* IdpPacket::Data ()
{
    return _buffer.get ();
}

std::shared_ptr<uint8_t> IdpPacket::Buffer ()
{
    return _buffer;
}

uint8_t* IdpPacket::Payload ()
{
    return _buffer.get () + 10;
}

uint8_t* IdpPacket::WritePointer ()
{
    return _buffer.get () + _writeIndex;
}

void IdpPacket::IncrementWritePointer (uint32_t length)
{
    _writeIndex += length;
}

uint32_t IdpPacket ::Length ()
{
    uint32_t result;

    memcpy (&result, (const void*) (Data () + 1), sizeof (uint32_t));

    if (BitConverter::IsLittleEndian ())
    {
        result = BitConverter::SwapEndian (result);
    }

    return result;
}

uint32_t IdpPacket::PayloadLength ()
{
    uint32_t length = Length () - 11;

    if (((uint8_t) Flags () & (uint8_t) IdpFlags::CRC) ==
        (uint8_t) IdpFlags::CRC)
    {
        length -= 4;
    }

    return length;
}

void IdpPacket::ResetRead ()
{
    _readIndex = 0;
}

void IdpPacket::ResetReadToPayload ()
{
    _readIndex = 10;
}

IdpPacket::~IdpPacket ()
{
}

File: src/cs/IdpProtocol/IdpPacket.cs
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for full license information.

using MiscUtil.Conversion;
using System;
using System.Runtime.InteropServices;

namespace IdpProtocol
{
    public enum IdpFlags
    {
        None = 0,
        CRC = 0x01
    }

    public class IdpPacket
    {
        private UInt32 _writeIndex;
        private UInt32 _length;
        private bool _isSealed;
        private int _readIndex;

        public IdpPacket(UInt32 payloadLength, IdpFlags flags, UInt16 source, UInt16 destination, bool isSealed = false)
        {
            _writeIndex = 0;
            _readIndex = 0;
            _isSealed = false;

            _length = 11 + payloadLength;

            if (flags.HasFlag(IdpFlags.CRC))
            {
                _length += 4;
            }

            Data = new byte[_length];

            Source = source;
            Destination = destination;

            Write((byte)0x02);
            Write(_length);
            Write((byte)flags);
            Write(source);
            Write(destination);

            _isSealed = isSealed;
        }

        public void Seal()
        {
            if (!_isSealed)
            {
                Write((byte)0x03);

                if (Flags.HasFlag(IdpFlags.CRC))
                {
                    //todo implement crc generation.
                    Write((UInt32)0xAA55AA55);
                }
            }

            _isSealed = true;
        }

        public void ResetRead()
        {
            _readIndex = 0;
        }

        public void ResetReadToPayload()
        {
            _readIndex = 10;
        }

        public T Read<T>() where T : struct
        {
            object result = null;
            bool increment = true;

            if (typeof(T) == typeof(bool))
            {
                result = EndianBitConverter.Big.ToBoolean(Data, _readIndex);

                increment = false;

                _readIndex++;
            }
            else if (typeof(T) == typeof(Int32))
            {
                result = EndianBitConverter.Big.ToInt32(Data, _readIndex);
            }
            else if (typeof(T) == typeof(UInt32))
            {
                result = EndianBitConverter.Big.ToUInt32(Data, _readIndex);
            }
            else if (typeof(T) == typeof(Int16))
            {
                result = EndianBitConverter.Big.ToInt16(Data, _readIndex);
            }
            else if (typeof(T) == typeof(UInt16))
            {
                result = EndianBitConverter.Big.ToUInt16(Data, _readIndex);
            }
            else if (typeof(T) == typeof(char))
            {
                result = EndianBitConverter.Big.ToChar(Data, _readIndex);
            }
            else if (typeof(T) == typeof(byte))
            {
                result = Data[_readIndex];
            }
            else if (typeof(T) == typeof(float))
            {
                result = EndianBitConverter.Big.ToSingle(Data, _readIndex);
            }
            else if (typeof(T) == typeof(IdpCommandFlags))
            {
                result = (IdpCommandFlags)Data[_readIndex];

                increment = false;

                _readIndex++;
            }
            else if (typeof(T) == typeof(IdpResponseCode))
            {
                result = (IdpResponseCode)Data[_readIndex];

                increment = false;

                _readIndex++;
            }
            else
            {
                throw new NotSupportedException($"Unable to decode type {typeof(T).FullName} from response.");
            }

            if (increment)
            {
                _readIndex += Marshal.SizeOf(result);
            }

            return (T)result;
        }

        public UInt16 Source { get; }

        public UInt16 Destination { get; }

        public byte[] Data { get; }

        public IdpFlags Flags => (IdpFlags)Data[5];

        public void Write(byte[] data)
        {
            Buffer.BlockCopy(data, 0, Data, (int)_writeIndex, data.Length);

            _writeIndex += (uint)data.Length;
        }

        public void Write<T>(T value) where T : struct
        {
            int size = Marshal.SizeOf(value);
            var bytes = new byte[size];
            IntPtr ptr = Marshal.AllocHGlobal(size);
            Marshal.StructureToPtr(value, ptr, true);
            Marshal.Copy(ptr, bytes, 0, size);
            Marshal.FreeHGlobal(ptr);

            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }

            Write(bytes);
        }
    }
}
